examples/c/test_hdf5.c

```diff
@@ -72,216 +91,485 @@ int main(int argc, char* argv[])
+        ret = PIOc_put_att(ncid, varid_time, "calendar", PIO_CHAR, strlen("noleap"), "noleap"); ERR
+        ret = PIOc_put_att(ncid, varid_time, "bounds", PIO_CHAR, strlen("time_bounds"), "time_bounds"); ERR
+
+        ret = PIOc_put_att(ncid, varid_lon, "long_name", PIO_CHAR, strlen("coordinate longitude"), "coordinate longitude"); ERR
+        ret = PIOc_put_att(ncid, varid_lon, "units", PIO_CHAR, strlen("degrees_east"), "degrees_east"); ERR
+        ret = PIOc_put_att(ncid, varid_lon, "_FillValue", PIO_FLOAT, 1, &att_fill_Value); ERR
+        ret = PIOc_put_att(ncid, varid_lon, "missing_value", PIO_FLOAT, 1, &att_missing_value); ERR
+
+        ret = PIOc_put_att(ncid, varid_lat, "long_name", PIO_CHAR, strlen("coordinate latitude"), "coordinate latitude"); ERR
+        ret = PIOc_put_att(ncid, varid_lat, "units", PIO_CHAR, strlen("degrees_north"), "degrees_north"); ERR
+        ret = PIOc_put_att(ncid, varid_lat, "_FillValue", PIO_FLOAT, 1, &att_fill_Value); ERR
+        ret = PIOc_put_att(ncid, varid_lat, "missing_value", PIO_FLOAT, 1, &att_missing_value); ERR
+
+        dimids[0] = dimid_time;
+        dimids[1] = dimid_lat;
+        dimids[2] = dimid_lon;
+        for (int i = 0; i < MAX_LAT_LON_VARS; i++)
+        {
+            snprintf(varname, PIO_MAX_NAME, "var_lat_lon_%03d", i);
+            ret = PIOc_def_var(ncid, varname, PIO_FLOAT, 3, dimids, &varid); ERR
+        }
+
+        ret = PIOc_enddef(ncid); ERR
+
+        ret = PIOc_closefile(ncid); ERR
+
+        if (my_rank == 0)
+        {
+            printf("Test writing %s with HDF5 IO type end\n", filename);
+            fflush(stdout);
+        }
+
+#ifdef _NETCDF4
+        /* Direct read support for HDF5 IO type is not implemented yet: SCORPIO implicitly switches HDF5 type to
+         * NETCDF4 type for reading back output files (generated by HDF5 type, fully NETCDF4 compatible). */
+        if (my_rank == 0)
+        {
+            printf("Test reading %s with NETCDF4 IO type start\n", filename);
+            fflush(stdout);
+        }
+
+        ncid = -1;
+        ret = PIOc_openfile(iosysid, &ncid, &format, filename, PIO_NOWRITE); ERR
+
+        ndims = -1;
+        ret = PIOc_inq_ndims(ncid, &ndims); ERR
+        assert(ndims == 3);
+
+        dimid_time = -1;
+        ret = PIOc_inq_dimid(ncid, "time", &dimid_time); ERR
+        assert(dimid_time >= 0 && dimid_time < ndims);
+
+        nvars = -1;
+        ret = PIOc_inq_nvars(ncid, &nvars); ERR
+        assert(nvars == (3 + MAX_LAT_LON_VARS));
+
+        ret = PIOc_closefile(ncid); ERR
+
+        if (my_rank == 0)
+        {
+            printf("Test reading %s with NETCDF4 IO type end\n", filename);
+            fflush(stdout);
+        }
+#endif
 
-    ret = PIOc_finalize(iosysid); ERR
+        ret = PIOc_finalize(iosysid); ERR
+    }
 
     MPI_Finalize();
 
 #ifdef TIMING
     GPTLfinalize();
 #endif
 
+#else
+    printf("SCORPIO is not configured with HDF5 support.\n");
```

Comment:
In a future PR, make the test fail if HDF5 support is not enabled/available. In cmake also ensure that this test is only included/run if direct HDF5 support is enabled.

Reply:
OK, thanks for the suggestion.